docs/proposals/formatted-output.md

```diff
@@ -0,0 +1,319 @@
+
+Fetch a specified layer and show the formatted JSON of the fetched manifest in formatted JSON.
+
+```shell
+oras manifest fetch $REGISTRY/$REPO:$TAG --format '{{(first .Layers).Reference}}'
+```
+
+```json
+{
+    "reference": "$REGISTRY/$REPO:$manifest_digest",
+    "layers" : [
+        {
+            "mediaType": "application/vnd.oci.image.layer.v1.tar",
+            "digest": "sha256:d2a84f4b8b650937ec8f73cd8be2c74add5a911ba64df27458ed8229da804a26",
+            "size": 12,
+            "annotations": {
+                "org.opencontainers.image.created":"2023-11-29T06:32:43Z"
+            },
+            "reference": "$REGISTRY/$REPO:$first_layer_digest",
+        },
+        {
+            "mediaType": "application/vnd.oci.image.layer.v1.tar",
+            "digest": "sha256:d2a84f4b8b650937ec8f73cd8be2c74add5a911ba64df27458ed8229da804a26",
+            "size": 12,
+            "annotations": {
+                "org.opencontainers.image.created":"2023-11-30T06:32:43Z"
+            },
+            "reference": "$REGISTRY/$REPO:$second_layer_digest",
+        }
+    ]
+}
+```
+
+## Desired user experience for proposal 2
+
+In order to  the manifest fields in the output, format the output using the given [Go template](https://golang.org/pkg/text/template/). The keys of the returned JSON can be used as the values for the `--format` flag.
+
+### Format the output using the given Go template
+
+For example, push an artifact to a repository and filter out the value of `reference` and `artifactType` of the pushed artifact in the standard output.
+
+```shell
+oras push $REGISTRY/$REPO:$TAG --format "{{.Reference}}', '{{.ArtifactType}}"
+```
+
+```console
+"localhost:5000@sha256:85438e6598bf35057962fff34399a362d469ca30a317939427fca6b7a289e70d, "application/vnd.example+type"  
+```
+
+For example, pull a file and filter out the specified fields `mediaType`, `reference`, `size` of the pulled file in the standard output.
+
+```shell
+oras pull $REGISTRY/$REPO:$TAG --format "{{.MediaType}}, {{.Reference}}, {{.Size}}"
+```
+
+```console
+"application/vnd.oci.image.layer.v1.tar","sha256:85438e6598bf35057962fff34399a362d469ca30a317939427fca6b7a289e70d", 12
+```
+
+For example, filter out the specified annotation value of an artifact by the key name `org.opencontainers.image.created`, with the [index function](https://pkg.go.dev/text/template#pkg-functions) defined in Go template.
+
+```shell
+oras discover localhost:5000/hello:v1 --format '{{index .Manifest.Annotations "org.opencontainers.image.created"}}'
+```
+
+```console
+"2023-11-29T06:32:43Z"
+```
+
+## FAQ
+
+- Why not consider extending the existing `--output` flag to enable JSON formatted output?
+
+`--output` has been used in other oras commands like `oras pull`, `oras manifest fetch` to output the file directory or file, it will be a breaking change if we extend the
+`--output` flag to enable JSON formatted output. 
+
+- Why choose [Go template](https://pkg.go.dev/text/template)?
```

Comment:
I understand the Go template approach but very often the problem people face with those templates is what is the data they can query for. The approach is good but I think we need to publish the schema for the outputs in our documentation.

Reply:
User can use `--format json` to easily view the schema. We should document this usage.

Reply:
We will document common functions provided by Go template to help users query data from the JSON output.